drivers/msgbox/sunxi-msgbox.c

```diff
@@ -7,139 +7,193 @@
+#define IRQ_STATUS_REG        0x0050
+#define RX_IRQ(n)             BIT(0 + 4 * (n))
+#define TX_IRQ(n)             BIT(3 + 4 * (n))
+
+/* IRQ information for the remote endpoint. */
+#define REMOTE_IRQ_EN_REG     0x0060
+#define REMOTE_IRQ_STATUS_REG 0x0070
+#define REMOTE_RX_IRQ(n)      BIT(2 + 4 * (n))
+#define REMOTE_TX_IRQ(n)      BIT(1 + 4 * (n))
 
-#define XMIT_FIFO_STATUS_REG(n) (0x0104 + 0x8 * (n))
-#define RECV_FIFO_STATUS_REG(n) (0x0100 + 0x8 * (n))
-#define FIFO_STATUS_MASK        BIT(0)
+#define RX_FIFO_STATUS_REG(n) (0x0100 + 0x8 * (n))
+#define TX_FIFO_STATUS_REG(n) (0x0104 + 0x8 * (n))
+#define FIFO_STATUS_MASK      BIT(0)
 
-#define XMIT_MSG_STATUS_REG(n)  (0x0144 + 0x8 * (n))
-#define RECV_MSG_STATUS_REG(n)  (0x0140 + 0x8 * (n))
-#define MSG_STATUS_MASK         BITMASK(0, 3)
+#define RX_MSG_STATUS_REG(n)  (0x0140 + 0x8 * (n))
+#define TX_MSG_STATUS_REG(n)  (0x0144 + 0x8 * (n))
+#define MSG_STATUS_MASK       BITMASK(0, 3)
 
-#define XMIT_MSG_DATA_REG(n)    (0x0184 + 0x8 * (n))
-#define RECV_MSG_DATA_REG(n)    (0x0180 + 0x8 * (n))
+#define RX_MSG_DATA_REG(n)    (0x0180 + 0x8 * (n))
+#define TX_MSG_DATA_REG(n)    (0x0184 + 0x8 * (n))
 
-static inline msg_handler
+static bool sunxi_msgbox_tx_pending(struct device *dev, uint8_t chan);
+
+static inline msgbox_handler
 get_handler(struct device *dev, uint8_t chan)
 {
-	return ((msg_handler *)dev->drvdata)[chan];
+	return ((msgbox_handler *)dev->drvdata)[chan];
 }
 
 static inline void
-set_handler(struct device *dev, uint8_t chan, msg_handler handler)
+set_handler(struct device *dev, uint8_t chan, msgbox_handler handler)
 {
-	((msg_handler *)dev->drvdata)[chan] = handler;
+	((msgbox_handler *)dev->drvdata)[chan] = handler;
 }
 
-static void
-sunxi_msgbox_handle_msg(struct device *dev, uint8_t chan)
+static bool
+sunxi_msgbox_peek_data(struct device *dev, uint8_t chan)
 {
-	msg_handler handler;
-	uint32_t    msg = mmio_read32(dev->regs + RECV_MSG_DATA_REG(chan));
+	uint32_t reg;
 
-	if ((handler = get_handler(dev, chan)))
-		handler(dev, chan, msg);
-	else
-		debug("%s: Unsolicited message %08x in channel %d",
-		      dev->name, msg, chan);
-}
-
-static void
-sunxi_msgbox_irq(struct device *dev)
-{
-	uint32_t reg = mmio_read32(dev->regs + IRQ_STATUS_REG);
+	assert(chan < SUNXI_MSGBOX_CHANS);
 
-	for (uint8_t chan = 0; chan < SUNXI_MSGBOX_CHANS; ++chan) {
-		if (!(reg & RECV_IRQ(chan)))
-			continue;
-		while (mmio_read32(dev->regs + RECV_MSG_STATUS_REG(chan)))
-			sunxi_msgbox_handle_msg(dev, chan);
-	}
+	reg = mmio_read32(dev->regs + RX_MSG_STATUS_REG(chan));
 
-	/* Clear all processed pending interrupts. */
-	mmio_write32(dev->regs + IRQ_STATUS_REG, reg);
+	return (reg & MSG_STATUS_MASK) > 0;
 }
 
 static int
 sunxi_msgbox_register_handler(struct device *dev, uint8_t chan,
-                              msg_handler handler)
+                              msgbox_handler handler)
 {
 	assert(chan < SUNXI_MSGBOX_CHANS);
-	assert(handler);
+	assert(handler != NULL);
```

Comment:
Same here: I think we can generate more helpful debug messages here.

Reply:
I think it should be clear that sending the driver a `NULL` function pointer is a bug. The assertion will print out the expression of the check that failed, so you'll still know exactly what caused the failure.